test/e2e/e2e.go

```diff
@@ -0,0 +1,234 @@
+		return err
+	}
+	cleaner.Add(rbacV1beta1.SchemeGroupVersion.Group, rbacV1beta1.SchemeGroupVersion.Version, "clusterrolebindings", "", res.ObjectMeta.Name)
+	return nil
+}
+
+// CreateServiceAccountAndBinding creates both ServiceAccount and ClusterRoleBinding with default
+// cluster-admin role
+func CreateServiceAccountAndBinding(clients *test.Clients, name string, logger *logging.BaseLogger, cleaner *test.Cleaner) error {
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: pkgTest.Flags.Namespace,
+		},
+	}
+	err := CreateServiceAccount(clients, sa, logger, cleaner)
+	if err != nil {
+		return err
+	}
+	crb := &rbacV1beta1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "e2e-tests-admin",
+		},
+		Subjects: []rbacV1beta1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      name,
+				Namespace: pkgTest.Flags.Namespace,
+			},
+		},
+		RoleRef: rbacV1beta1.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+	err = CreateClusterRoleBinding(clients, crb, logger, cleaner)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreatePod will create a Pod
+func CreatePod(clients *test.Clients, pod *corev1.Pod, logger *logging.BaseLogger, cleaner *test.Cleaner) error {
+	pods := clients.Kube.Kube.CoreV1().Pods(pod.GetNamespace())
+	res, err := pods.Create(pod)
+	if err != nil {
+		return err
+	}
+	cleaner.Add(corev1.SchemeGroupVersion.Group, corev1.SchemeGroupVersion.Version, "pods", res.ObjectMeta.Namespace, res.ObjectMeta.Name)
+	return nil
+}
+
+// PodLogs returns Pod logs for given Pod and Container
+func PodLogs(clients *test.Clients, podName string, containerName string, logger *logging.BaseLogger) ([]byte, error) {
+	pods := clients.Kube.Kube.CoreV1().Pods(pkgTest.Flags.Namespace)
+	podList, err := pods.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		if strings.Contains(pod.Name, podName) {
+			result := pods.GetLogs(pod.Name, &corev1.PodLogOptions{
+				Container: containerName,
+			}).Do()
+			return result.Raw()
+		}
+	}
+	return nil, fmt.Errorf("Could not find logs for %s/%s", podName, containerName)
+}
+
+// WaitForLogContent waits until logs for given Pod/Container include the given content.
+// If the content is not present within timeout it returns error.
+func WaitForLogContent(clients *test.Clients, logger *logging.BaseLogger, podName string, containerName string, content string) error {
+	return wait.PollImmediate(interval, timeout, func() (bool, error) {
+		logs, err := PodLogs(clients, podName, containerName, logger)
+		if err != nil {
+			return true, err
+		}
+		return strings.Contains(string(logs), content), nil
+	})
+}
+
+// WaitForAllPodsRunning will wait until all pods in the given namespace are running
+func WaitForAllPodsRunning(clients *test.Clients, logger *logging.BaseLogger, namespace string) error {
```

Comment:
Does this block forever or does it timeout at some point? Or put another way, if something goes wrong, will things block forevs?

Reply:
This timeouts after 8 minutes - the `WaitForPodListState` method called here is in knative/pkg and the timeout is set there.